cranelift/codegen/src/isa/s390x/inst.isle

```diff
@@ -3424,17 +3427,17 @@
             ;; Emit initial load followed by compare-and-swap loop.
             (_ Unit (emit_load (ty_ext32 ty) (casloop_val_reg) aligned_mem))
             (_ Unit (emit_loop ib (intcc_as_cond (IntCC.NotEqual)))))
-        result))
+        (preg_gpr_0)))
 
 ;; Compute the previous memory value after a (fullword) compare-and-swap loop.
 ;; In the big-endian case, the value is already correct, but may need to be
 ;; copied out of the hard register.  In the little-endian case, we need to
 ;; byte-swap since the compare-and-swap instruction is always big-endian.
-(decl casloop_result (Type MemFlags Reg) Reg)
+(decl casloop_result (Type MemFlags PReg) Reg)
 (rule 1 (casloop_result (ty_32_or_64 ty) (bigendian) result)
-      (copy_reg ty result))
+      (mov_preg result))
 (rule (casloop_result (ty_32_or_64 ty) (littleendian) result)
-      (bswap_reg ty result))
+      (bswap_reg ty (mov_preg result)))
```

Comment:
I'm confused about the exact rules when/which instructions are allowed to refer to non-allocatable fixed registers.  It seems the load before loop (loading into %r0) is OK, but the move is problematic (which I understand is because regalloc handles moves specially).

But is it OK to have a fixed register as source of the bswap instruction?  If so, we wouldn't have to incur this performance regression.  In fact, if *only* moves are the problem, maybe the fix could be as simple as having a special-cased rule for `copy_reg` that detects that the source is a fixed register and uses mov_preg instead?

Reply:
That's a good point: moves are indeed the case that we need to avoid, and it should be fine for the bswap instruction to take a fixed non-allocatable input. I'll fix this up, thanks!